pkgs/firehose/lib/firehose.dart

```diff
@@ -282,6 +275,187 @@ Documentation at https://github.com/dart-lang/ecosystem/wiki/Publishing-automati
+  }
+
+  Future<List<String>> _getFilesWithoutLicenses(Github github) async {
+    var dir = Directory.current;
+    var dartFiles = await dir
+        .list(recursive: true)
+        .where((f) => f.path.endsWith('.dart'))
+        .toList();
+    var filesWithoutLicenses = dartFiles
+        .map((file) {
+          var fileContents = File(file.path).readAsStringSync();
+          var fileContainsCopyright = fileContents.contains('// Copyright (c)');
+          if (!fileContainsCopyright) {
+            return path.relative(file.path, from: Directory.current.path);
+          }
+        })
+        .whereType<String>()
+        .toList();
+    return filesWithoutLicenses;
+  }
+
+  Future<void> createOrUpdateComment(
+    Github github,
+    int? existingCommentId,
+    String commentText,
+  ) async {
+    if (existingCommentId == null) {
+      await allowFailure(
+        github.createComment(
+            github.repoSlug!, github.issueNumber!, commentText),
+        logError: print,
+      );
+    } else {
+      await allowFailure(
+        github.updateComment(github.repoSlug!, existingCommentId, commentText),
+        logError: print,
+      );
+    }
+  }
+
+  Future<void> changelog() async {
+    var github = Github();
+
+    // Do basic validation of our expected env var.
+    if (!_expectEnv(github.githubAuthToken, 'GITHUB_TOKEN')) return;
+    if (!_expectEnv(github.repoSlug, 'GITHUB_REPOSITORY')) return;
+    if (!_expectEnv(github.issueNumber, 'ISSUE_NUMBER')) return;
+    if (!_expectEnv(github.sha, 'GITHUB_SHA')) return;
+
+    if ((github.actor ?? '').endsWith(_botSuffix)) {
+      print('Skipping license check for ${github.actor} PRs.');
+      return;
+    }
+
+    var filePaths = await _packagesWithoutChangelog(github);
+
+    String markdownResult;
+
+    if (filePaths.isNotEmpty) {
+      if (Repository().isSinglePackageRepo) {
+        markdownResult = '''
+Changes to the files ${filePaths.values.first.map((e) => path.relative(e, from: Directory.current.path)).join(', ')} need to be accounted for in the changelog''';
+      } else {
+        markdownResult = '''
+Changes to the files in need to be accounted for in the changelog:
+
+${filePaths.entries.map((e) => '${e.key.name}: ${e.value.map((e) => path.relative(e, from: Directory.current.path)).join(', ')}').join('\n')}
+''';
+      }
+    } else {
+      markdownResult = 'No unrecorded changes found.';
+    }
+
+    // Write the publish info status to the job summary.
+    github.appendStepSummary(markdownResult);
+
+    if (filePaths.isNotEmpty) {
+      var existingCommentId = await allowFailure(
+        github.findCommentId(
+          github.repoSlug!,
+          github.issueNumber!,
+          user: _githubActionsUser,
+          searchTerm: _changelogBotTag,
+        ),
+        logError: print,
+      );
+
+      var commentText = '$_changelogBotTag\n\n$markdownResult';
+
+      await createOrUpdateComment(github, existingCommentId, commentText);
```

Comment:
FYI that this workflow won't be able to create comments on PRs that are from forks of the repo. So we'll have less (easy) visibility of issues there. We can still fail the workflow - which they'll see - and write the 'step summary' content below.

For forks, I think the most visibility will be the workflow failure and then the text output of the workflow once they dig into that. The comment is very visual and nice, and we should retain it for the PRs that aren't from forks.